### PR TITLE
feat(terraform): nested modules support in graph

### PR DESCRIPTION
**Pull request by achiar99**
**By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.**

[//]: # "
    # PR Title
    Be aware that we use the title to create changelog automatically and therefore only allow specific prefixes
    - break:    to indicate a breaking change, this supersedes any of the types
    - feat:     to indicate new features or checks
    - fix:      to indicate a bugfix or handling of edge cases of existing checks
    - docs:     to indicate an update to our documentation
    - chore:    to indicate adjustments to workflow files or dependency updates
    - platform: to indicate a change needed for the platform
    Additionally a scope is needs to be added to the prefix, which indicates the targeted framework, in doubt choose 'general'.
    ex.
    feat(terraform): add CKV_AWS_123 to ensure that VPC Endpoint Service is configured for Manual Acceptance
"

## Description

Support nested module instances when building tf graph
This feature locked with env var `CHECKOV_ENABLE_NESTED_MODULES`

1) change the tf definition creation to support multiple nested modules and resources
2) change the dependencies map calculation logic (this solve bugs in the current graph creation)
3) change breadcrumbs keys to new nested modules keys
4) change dependency path format (`a->b->c`) to `nested_nested[nested#0[main#1]]`
5) duplicate tests that fail with the new logic

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my feature, policy, or fix is effective and works
- [ ] New and existing tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: checkov/terraform/parser.py [hunk 1]
```diff
@@ -391,9 +398,20 @@ def _load_modules(self, root_dir: str, module_loader_registry: ModuleLoaderRegis
...
 
+                    if self.enable_nested_modules:
+                        file_key = self.get_file_key(file, nested_modules_data)
+                        current_nested_data = (file_key, module_index, module_call_name)
+
+                        if current_nested_data in self.module_to_resolved:
+                            resolved_loc_list = self.module_to_resolved[current_nested_data]
+                        else:
+                            resolved_loc_list = []
+                            self.module_to_resolved[current_nested_data] = resolved_loc_list
```
Comment (tronxd) [inline checkov/terraform/parser.py:409]:
```suggestion
                        resolved_loc_list = []
                        if current_nested_data in self.module_to_resolved:
                            resolved_loc_list = self.module_to_resolved[current_nested_data]
                        self.module_to_resolved[current_nested_data] = resolved_loc_list
```

File: checkov/terraform/parser.py [hunk 2]
```diff
@@ -522,6 +555,15 @@ def parse_hcl_module(
 
         return module, tf_definitions
 
+    def _update_resolved_modules(self):
+        for key, resolved_list in self.module_to_resolved.items():
+            file_key = key[0]
+            module_index = key[1]
+            module_name = key[2]
```
Comment (tronxd) [inline checkov/terraform/parser.py:562]:
```suggestion
            file_key, module_index, module_name = key
```

File: checkov/terraform/parser.py [hunk 3]
```diff
@@ -577,6 +619,21 @@ def _clean_parser_types(conf: dict[str, Any]) -> dict[str, Any]:
                 sorted_conf[attribute] = str(values)
         return sorted_conf
 
+    def get_file_key(self, file, nested_data):
+        if not nested_data:
+            return f'{file}'
+        nested_str = self.get_file_key(nested_data.get("file"), nested_data.get('nested_modules_data'))
+        nested = f'{file}[{nested_str}#{nested_data.get("module_index")}]'
```
Comment (tronxd) [inline checkov/terraform/parser.py:626]: @achiar99 Can we refactor the path creation logic into a function? (`{key}[file]#{module_index}`)

File: checkov/terraform/parser.py [hunk 4]
```diff
@@ -86,6 +86,9 @@ def _init(self, directory: str, out_definitions: Optional[Dict],
         if self.env_vars is None:
             self.env_vars = dict(os.environ)
         self.excluded_paths = excluded_paths
+        self.enable_nested_modules = os.getenv('ENABLE_NESTED_MODULES', 'False') == 'True'
```
Comment (gruebel) [inline checkov/terraform/parser.py:89]:
```suggestion
        self.enable_nested_modules = os.getenv('CHECKOV_ENABLE_NESTED_MODULES', 'False') == 'True'
```

File: checkov/terraform/parser.py [hunk 5]
```diff
@@ -577,6 +619,21 @@ def _clean_parser_types(conf: dict[str, Any]) -> dict[str, Any]:
                 sorted_conf[attribute] = str(values)
         return sorted_conf
 
+    def get_file_key(self, file, nested_data):
+        if not nested_data:
+            return f'{file}'
```
Comment (gruebel) [inline checkov/terraform/parser.py:624]:
```suggestion
            return str(file)
```

**Review by gruebel — COMMENTED** [review]
nice work, added some extra comments next Naors 🙂

**Comment by nimrodkor** [comment]
@achiar99 Please merge from `main`

**Review by tronxd — APPROVED** [review]

File: checkov/terraform/parser.py [hunk 6]
```diff
@@ -577,6 +625,27 @@ def _clean_parser_types(conf: dict[str, Any]) -> dict[str, Any]:
...
+    def get_new_nested_module_key(self, key, file, module_index, nested_data) -> str:
+        if not nested_data:
+            return get_tf_definition_key(key, file, module_index)
+        visited_key_to_add = get_tf_definition_key(key, file, module_index)
+        self.visited_definition_keys.add(visited_key_to_add)
+        nested_key = self.get_new_nested_module_key('', nested_data.get('file'),
+                                                    nested_data.get('module_index'),
+                                                    nested_data.get('nested_modules_data'))
+        new_key = get_tf_definition_key(key, file, module_index, nested_key)
+        return new_key
```
Comment (gruebel) [inline checkov/terraform/parser.py:647]:
```suggestion
        return get_tf_definition_key(key, file, module_index, nested_key)
```
to align it with the method above 😄

File: checkov/terraform/parser.py [hunk 7]
```diff
@@ -642,6 +711,36 @@ def get_next_vertices(evaluated_files: list[str], unevaluated_files: list[str])
...
+    @staticmethod
+    def get_nested_modules_data_as_list(file_path):
+        path = get_abs_path(file_path)
+        modules_list = []
+
+        while is_nested(file_path):
+            module, index = get_module_from_full_path(file_path)
+            modules_list.append((module, index))
+            file_path = module
+        return list(reversed(modules_list)), path
```
Comment (gruebel) [inline checkov/terraform/parser.py:723]:
```suggestion
            file_path = module
        modules_list.reverse()
        return modules_list, path
```
no need to create a new list

**Review by gruebel — COMMENTED** [review]
nice work so far 💪

File: checkov/terraform/runner.py [hunk 8]
```diff
@@ -403,8 +440,11 @@ def run_block(self, entities,
                         details=check.details
                     )
                     if CHECKOV_CREATE_GRAPH:
-                        breadcrumb = self.breadcrumbs.get(record.file_path, {}).get(
-                            '.'.join([entity_type, entity_name]))
+                        if self.enable_nested_modules:
+                            entity_key = entity_id
+                        else:
+                            entity_key = '.'.join([entity_type, entity_name])
```
Comment (gruebel) [inline checkov/terraform/runner.py:446]:
```suggestion
                            entity_key = f"{entity_type}.{entity_name}"
```

**Review by gruebel — APPROVED** [review]
noice 🎄

**Pull request merged**